src/Core/src/AwsError/AwsErrorFactory.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace AsyncAws\Core\AwsError;
+
+use AsyncAws\Core\Exception\ParseResponse;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+class AwsErrorFactory
+{
+    public static function createFromResponse(ResponseInterface $response): AwsError
+    {
+        $content = $response->getContent(false);
+        $headers = $response->getHeaders(false);
+
+        return self::createFromContent($content, $headers);
+    }
+
+    public static function createFromContent(string $content, array $headers): AwsError
+    {
+        $e = null;
+
+        try {
+            // Try json_decode it first, fallback to XML
+            if ($body = json_decode($content, true)) {
+                return self::parseJson($body, $headers);
+            }
+
+            set_error_handler(
+                static function ($errno, $errstr, $errfile, $errline) {
+                    throw new \RuntimeException($errstr, $errno);
+                }
+            );
+
+            try {
+                $xml = new \SimpleXMLElement($content);
+            } finally {
+                restore_error_handler();
+            }
+
+            return self::parseXml($xml);
+        } catch (\Throwable $e) {
+        }
+
+        throw new ParseResponse('Failed to parse AWS error: ' . $content, 0, $e);
```

Comment:
Why not adding this inside the `catch`?

Reply:
good catch. I forgot that try always returns